Software/Program/Code/Core/Src/main.c

```diff
@@ -40,24 +41,80 @@
 /* USER CODE END PM */
 
 /* Private variables ---------------------------------------------------------*/
+ADC_HandleTypeDef hadc3;
 
 TIM_HandleTypeDef htim2;
 
 /* USER CODE BEGIN PV */
 
+/* Public variables ---------------------------------------------------------*/
+static volatile bool OnHighPulse;
+static volatile bool ButtonPressed;
+static volatile bool ADC_ConversionComplete;
+static uint32_t ARR_Reload_Values[] = { ARR_TABLE_INITIALIZATION_VALUE, ARR_TABLE_INITIALIZATION_VALUE };
+static uint8_t ARR_Reload_Value_Idx;
+
 /* USER CODE END PV */
 
 /* Private function prototypes -----------------------------------------------*/
 void SystemClock_Config(void);
 static void MX_GPIO_Init(void);
 static void MX_TIM2_Init(void);
+static void MX_ADC3_Init(void);
 /* USER CODE BEGIN PFP */
 
 /* USER CODE END PFP */
 
 /* Private user code ---------------------------------------------------------*/
 /* USER CODE BEGIN 0 */
 
+void HAL_TIM_PeriodElapsedCallback( TIM_HandleTypeDef * timer_handle )
+{
+   if ( timer_handle == &htim2 )
+   {
+      HAL_GPIO_TogglePin( FRQ_OUT_PORT, FRQ_OUT_PIN );
+      
+      // Load next ARR value into ARR
+      __HAL_TIM_SET_AUTORELOAD( &htim2, ARR_Reload_Values[ ARR_Reload_Value_Idx ] );
+      
+      // Update index for next run
+      ARR_Reload_Value_Idx = ( ARR_Reload_Value_Idx + 1 ) % (sizeof(ARR_Reload_Values) / sizeof(uint32_t));
```

Comment:
```suggestion
      ARR_Reload_Value_Idx = ( ARR_Reload_Value_Idx ^ 0x01u ) & 0x01u;
```
Since the `ARR_Reload_Values` array really only needs to be of length 2, you can avoid the divide operation with an XOR `⊕` and an AND `&`.